### ISSUE TITLE
Redis cluster mode returns: "panic: cannot decode redis array into string"

### ISSUE THREAD

**Issue by renuka-fernando** [envoyproxy/ratelimit#351]
### Sample Redis cluster

Full docker compose setup wtih a sample Redis cluster: [redis-cluster-docker-compose.tar.gz](https://github.com/envoyproxy/ratelimit/files/9348680/redis-cluster-docker-compose.tar.gz)

Start setup with

```sh
tar -xzf redis-cluster-docker-compose.tar.gz
cd redis-cluster-docker-compose/
docker-compose up -d; docker-compose logs -f redis_cli ratelimit
```

#### Logs

```logs
redis_cli_1  | Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe.
redis_cli_1  | >>> Performing hash slots allocation on 3 nodes...
redis_cli_1  | Master[0] -> Slots 0 - 5460
redis_cli_1  | Master[1] -> Slots 5461 - 10922
redis_cli_1  | Master[2] -> Slots 10923 - 16383
redis_cli_1  | M: 8a8621d19057d63cdf8fd5c33fe7a7ad3371fc19 172.20.0.30:7000
redis_cli_1  |    slots:[0-5460] (5461 slots) master
redis_cli_1  | M: f0dcff6c2b8d3382636010f1987c1e481557d737 172.20.0.31:7001
redis_cli_1  |    slots:[5461-10922] (5462 slots) master
redis_cli_1  | M: d35706e2998fc889f16f6d58883877532c0f22ce 172.20.0.32:7002
redis_cli_1  |    slots:[10923-16383] (5461 slots) master
redis_cli_1  | >>> Nodes configuration updated
redis_cli_1  | >>> Assign a different config epoch to each node
redis_cli_1  | >>> Sending CLUSTER MEET messages to join the cluster
redis_cli_1  | Waiting for the cluster to join
redis_cli_1  |
redis_cli_1  | >>> Performing Cluster Check (using node 172.20.0.30:7000)
redis_cli_1  | M: 8a8621d19057d63cdf8fd5c33fe7a7ad3371fc19 172.20.0.30:7000
redis_cli_1  |    slots:[0-5460] (5461 slots) master
redis_cli_1  | M: d35706e2998fc889f16f6d58883877532c0f22ce 172.20.0.32:7002
redis_cli_1  |    slots:[10923-16383] (5461 slots) master
redis_cli_1  | M: f0dcff6c2b8d3382636010f1987c1e481557d737 172.20.0.31:7001
redis_cli_1  |    slots:[5461-10922] (5462 slots) master
redis_cli_1  | [OK] All nodes agree about slots configuration.
redis_cli_1  | >>> Check for open slots...
redis_cli_1  | >>> Check slots coverage...
redis_cli_1  | [OK] All 16384 slots covered.
redis_cli_1  | Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe.
redis_cli_1  | 172.20.0.30:7000 (8a8621d1...) -> 0 keys | 5461 slots | 0 slaves.
redis_cli_1  | 172.20.0.32:7002 (d35706e2...) -> 0 keys | 5461 slots | 0 slaves.
redis_cli_1  | 172.20.0.31:7001 (f0dcff6c...) -> 0 keys | 5462 slots | 0 slaves.
redis_cli_1  | [OK] 0 keys in 3 masters.
redis_cli_1  | 0.00 keys per slot on average.
redis_cli_1  | >>> Performing Cluster Check (using node 172.20.0.30:7000)
redis_cli_1  | M: 8a8621d19057d63cdf8fd5c33fe7a7ad3371fc19 172.20.0.30:7000
redis_cli_1  |    slots:[0-5460] (5461 slots) master
redis_cli_1  | M: d35706e2998fc889f16f6d58883877532c0f22ce 172.20.0.32:7002
redis_cli_1  |    slots:[10923-16383] (5461 slots) master
redis_cli_1  | M: f0dcff6c2b8d3382636010f1987c1e481557d737 172.20.0.31:7001
redis_cli_1  |    slots:[5461-10922] (5462 slots) master
redis_cli_1  | [OK] All nodes agree about slots configuration.
redis_cli_1  | >>> Check for open slots...
redis_cli_1  | >>> Check slots coverage...
redis_cli_1  | [OK] All 16384 slots covered.
redis-cluster-docker-compose_redis_cli_1 exited with code 0

ratelimit_1  | starting ratelimiter now
ratelimit_1  | time="2022-08-16T07:26:33Z" level=warning msg="statsd is not in use"
ratelimit_1  | time="2022-08-16T07:26:33Z" level=info msg="Tracing disabled"
ratelimit_1  | time="2022-08-16T07:26:33Z" level=debug msg="runtime changed. loading new snapshot at /data/ratelimit/config"
ratelimit_1  | time="2022-08-16T07:26:33Z" level=debug msg="runtime: processing /data/ratelimit/config"
ratelimit_1  | time="2022-08-16T07:26:33Z" level=debug msg="runtime: processing /data/ratelimit/config/config.yaml"
ratelimit_1  | time="2022-08-16T07:26:33Z" level=debug msg="runtime: adding key=config.yaml value=---\ndomain: mongo_cps\ndescriptors:\n  - key: database\n    value: users\n    rate_limit:\n      unit: second\n      requests_per_unit: 500\n\n  - key: database\n    value: default\n    rate_limit:\n      unit: second\n      requests_per_unit: 500\n uint=false"
ratelimit_1  | time="2022-08-16T07:26:33Z" level=debug msg="runtime: processing /data/ratelimit/config/example.yaml"
ratelimit_1  | time="2022-08-16T07:26:33Z" level=debug msg="runtime: adding key=example.yaml value=---\ndomain: rl # we can have org level domains, then we have to configure router level rate limit filters\ndescriptors:\n  - key: k1\n    value: v1\n    rate_limit:\n      unit: minute\n      requests_per_unit: 2\n    descriptors:\n      - key: k2\n        # value: v1\n        rate_limit:\n          unit: minute\n          requests_per_unit: 12\n  - key: k2\n    # value: v1\n    rate_limit:\n      unit: minute\n      requests_per_unit: 16\n  - key: source_cluster\n    value: proxy\n    descriptors:\n      - key: destination_cluster\n        value: mock\n        rate_limit:\n          unit: minute\n          requests_per_unit: 1\n  - key: cl\n    rate_limit:\n      unit: minute\n      requests_per_unit: 2\n    descriptors:\n      - key: bar\n        rate_limit:\n          unit: minute\n          requests_per_unit: 5\n  - key: foo\n    rate_limit:\n      unit: minute\n      requests_per_unit: 2\n    descriptors:\n      - key: bar\n        rate_limit:\n          unit: minute\n          requests_per_unit: 3\n      - key: bar\n        value: banned\n        rate_limit:\n          unit: minute\n          requests_per_unit: 0\n      - key: baz\n        rate_limit:\n          unit: second\n          requests_per_unit: 1\n      - key: baz\n        value: not-so-shady\n        rate_limit:\n          unit: minute\n          requests_per_unit: 3\n      - key: baz\n        value: shady\n        rate_limit:\n          unit: minute\n          requests_per_unit: 3\n        shadow_mode: true\n      - key: bay\n        rate_limit:\n          unlimited: true\n  - key: qux\n    rate_limit:\n      unlimited: true\n uint=false"
ratelimit_1  | time="2022-08-16T07:26:33Z" level=warning msg="connecting to redis on redis_0:7000,redis_1:7001,redis_2:7002 with pool size 10"
ratelimit_1  | time="2022-08-16T07:26:33Z" level=debug msg="Implicit pipelining enabled: true"
ratelimit_1  | time="2022-08-16T07:26:33Z" level=warning msg="Creating cluster with urls [redis_0:7000 redis_1:7001 redis_2:7002]"
ratelimit_1  | time="2022-08-16T07:26:33Z" level=warning msg="enabling authentication to redis on redis_0:7000,redis_1:7001,redis_2:7002"
ratelimit_1  | time="2022-08-16T07:26:33Z" level=warning msg="enabling authentication to redis on redis_0:7000,redis_1:7001,redis_2:7002"
ratelimit_1  | panic: cannot decode redis array into string
ratelimit_1  |
ratelimit_1  | goroutine 1 [running]:
ratelimit_1  | github.com/envoyproxy/ratelimit/src/redis.checkError(...)
ratelimit_1  | 	/ratelimit/src/redis/driver_impl.go:63
ratelimit_1  | github.com/envoyproxy/ratelimit/src/redis.NewClientImpl({0xd493f0, 0xc00013d8e0}, 0x0, {0xc000036083, 0xb}, {0xc00003602a, 0x3}, {0xc00003609b, 0x7}, {0xc00003c00a, ...}, ...)
ratelimit_1  | 	/ratelimit/src/redis/driver_impl.go:126 +0xa74
ratelimit_1  | github.com/envoyproxy/ratelimit/src/redis.NewRateLimiterCacheImplFromSettings({0x0, {0xacaaed, 0x7}, 0x1f90, {0xad36a4, 0x7}, 0x17b6, {0xad215d, 0x7}, 0x1f91, ...}, ...)
ratelimit_1  | 	/ratelimit/src/redis/cache_impl.go:22 +0x2bd
ratelimit_1  | github.com/envoyproxy/ratelimit/src/service_cmd/runner.createLimiter({_, _}, {0x0, {0xacaaed, 0x7}, 0x1f90, {0xad36a4, 0x7}, 0x17b6, {0xad215d, ...}, ...}, ...)
ratelimit_1  | 	/ratelimit/src/service_cmd/runner/runner.go:55 +0x277
ratelimit_1  | github.com/envoyproxy/ratelimit/src/service_cmd/runner.(*Runner).Run(0xc0003faa80)
ratelimit_1  | 	/ratelimit/src/service_cmd/runner/runner.go:121 +0x595
ratelimit_1  | main.main()
ratelimit_1  | 	/ratelimit/src/service_cmd/main.go:10 +0x125
redis-cluster-docker-compose_ratelimit_1 exited with code 2
```

#### Docker Compose YAML file inside the tar file

```yaml
version: '3'
services:
  ratelimit:
    image: envoyproxy/ratelimit:master
    command: ["sh", "-c", "sleep 15; echo 'starting ratelimiter now'; /bin/ratelimit;"]
    volumes:
      - ./ratelimit/config:/data/ratelimit/config
    environment:
      - USE_STATSD=false
      - LOG_LEVEL=debug
      - REDIS_SOCKET_TYPE=tcp
      - REDIS_TYPE=cluster
      - REDIS_AUTH=password123
      - REDIS_URL=redis_0:7000,redis_1:7001,redis_2:7002
      - REDIS_PIPELINE_LIMIT=8
      - RUNTIME_ROOT=/data
      - RUNTIME_SUBDIRECTORY=ratelimit
      - RUNTIME_WATCH_ROOT=false
    depends_on:
      - redis_cli
      - redis_0
      - redis_1
      - redis_2
    networks:
      redis_subnet:
        ipv4_address: 172.20.0.10
  redis_cli:
    image: redis:7.0.4
    command:
      - "sh"
      - "-c"
      - "redis-cli --cluster create -a password123 172.20.0.30:7000 172.20.0.31:7001 172.20.0.32:7002 --cluster-replicas 0 --cluster-yes;    redis-cli --cluster check -a password123 172.20.0.30:7000;"
    depends_on:
      - redis_0
      - redis_1
      - redis_2
    networks:
      redis_subnet:
        ipv4_address: 172.20.0.20
  redis_0:
    image: redis:7.0.4
    ports:
      - 7000:7000
    volumes:
      - ./7000:/etc/redis-data
    command: ["redis-server", "/etc/redis-data/redis.conf"]
    networks:
      redis_subnet:
        ipv4_address: 172.20.0.30
  redis_1:
    image: redis:7.0.4
    ports:
      - 7001:7001
    volumes:
      - ./7001:/etc/redis-data
    networks:
      redis_subnet:
        ipv4_address: 172.20.0.31
    command: ["redis-server", "/etc/redis-data/redis.conf"]
  redis_2:
    image: redis:7.0.4
    ports:
      - 7002:7002
    volumes:
      - ./7002:/etc/redis-data
    command: ["redis-server", "/etc/redis-data/redis.conf"]
    networks:
      redis_subnet:
        ipv4_address: 172.20.0.32

networks:
  redis_subnet:
    driver: bridge
    ipam:
      config:
        - subnet: 172.20.0.0/24
```

**Comment by renuka-fernando** [comment]
I even tested with bintami helm chart https://github.com/bitnami/charts/tree/master/bitnami/redis-cluster
Still the same behavior.

Ratelimit server commit: 8655817d2bb155bf46b083876ffa0f74d9b3335b

#### Redis Cluster

```
helm install my-redis bitnami/redis-cluster --set global.redis.password=password123
```

#### Ratelimit server

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ratelimit
spec:
  replicas: 3
  selector:
    matchLabels:
      app: ratelimit
  template:
    metadata:
      labels:
        app: ratelimit
      annotations: {}
    spec:
      containers:
        - name: ratelimit
          image:envoyproxy/ratelimit:master
          imagePullPolicy: IfNotPresent
          command:
            - "/bin/ratelimit"
          env:
            - name: USE_STATSD
              value: "false" # TODO: false
            - name: LOG_LEVEL
              value: debug
            - name: REDIS_SOCKET_TYPE
              value: tcp
            - name: REDIS_TYPE
              value: cluster
            - name: REDIS_URL
              value: "my-redis-redis-cluster:6379"
            - name: REDIS_PIPELINE_LIMIT
              value: "8"
            - name: REDIS_AUTH
              value: "password123"
            - name: RUNTIME_ROOT
              value: /data
            - name: RUNTIME_SUBDIRECTORY
              value: ratelimit
            - name: RUNTIME_WATCH_ROOT
              value: "false"
          volumeMounts:
            - mountPath: /data/ratelimit/config
              name: ratelimit-config-vol
          resources:
            requests:
              memory: "500Mi"
              cpu: "500m"
            limits:
              memory: "500Mi"
              cpu: "500m"
          ports:
            - containerPort: 8081 # gRPC
            - containerPort: 8080 # HTTP JSON
            - containerPort: 6070 # debug
          livenessProbe:
            httpGet:
              port: 8080
              path: /healthcheck
              scheme: HTTP
            failureThreshold: 30
            periodSeconds: 5
          readinessProbe:
            httpGet:
              port: 8080
              path: /healthcheck
              scheme: HTTP
            failureThreshold: 30
            periodSeconds: 5
      volumes:
        - name: ratelimit-config-vol
          configMap:
            name: ratelimit-config
      restartPolicy: Always
```

#### Logs
```logs
$ k logs ratelimit-5759c7f8d4-p6bs7
time="2022-08-16T11:52:56Z" level=warning msg="statsd is not in use"
time="2022-08-16T11:52:56Z" level=info msg="Tracing disabled"
time="2022-08-16T11:52:56Z" level=debug msg="runtime changed. loading new snapshot at /data/ratelimit/config"
time="2022-08-16T11:52:56Z" level=debug msg="runtime: processing /data/ratelimit/config"
time="2022-08-16T11:52:56Z" level=debug msg="runtime: processing /data/ratelimit/config/..2022_08_16_11_52_54.4272888654"
time="2022-08-16T11:52:56Z" level=debug msg="runtime: processing /data/ratelimit/config/..2022_08_16_11_52_54.4272888654/ratelimit-config.yaml"
...

time="2022-08-16T11:52:56Z" level=debug msg="runtime: processing /data/ratelimit/config/..data"
time="2022-08-16T11:52:56Z" level=warning msg="runtime: error reading /data/ratelimit/config/..data: read /data/ratelimit/config/..data: is a directory"
time="2022-08-16T11:52:56Z" level=debug msg="runtime: processing /data/ratelimit/config/ratelimit-config.yaml"
...


time="2022-08-16T11:52:56Z" level=warning msg="connecting to redis on my-redis-redis-cluster:6379 with pool size 10"
time="2022-08-16T11:52:56Z" level=debug msg="Implicit pipelining enabled: true"
time="2022-08-16T11:52:56Z" level=warning msg="Creating cluster with urls [my-redis-redis-cluster:6379]"
time="2022-08-16T11:52:56Z" level=warning msg="enabling authentication to redis on my-redis-redis-cluster:6379"
time="2022-08-16T11:52:56Z" level=warning msg="enabling authentication to redis on my-redis-redis-cluster:6379"
panic: cannot decode redis array into string

goroutine 1 [running]:
github.com/envoyproxy/ratelimit/src/redis.checkError(...)
	/Users/renuka/git/ratelimit/src/redis/driver_impl.go:62
github.com/envoyproxy/ratelimit/src/redis.NewClientImpl({0xd481b0, 0xc0000cf8e0}, 0x0, {0xc00003a00b, 0xb}, {0xc00003a05a, 0x3}, {0xc00003a06b, 0x7}, {0xc00003603a, ...}, ...)
	/Users/renuka/git/ratelimit/src/redis/driver_impl.go:124 +0xa66
github.com/envoyproxy/ratelimit/src/redis.NewRateLimiterCacheImplFromSettings({0x0, {0xac9aaf, 0x7}, 0x1f90, {0xad2666, 0x7}, 0x17b6, {0xad111f, 0x7}, 0x1f91, ...}, ...)
	/Users/renuka/git/ratelimit/src/redis/cache_impl.go:22 +0x2bd
github.com/envoyproxy/ratelimit/src/service_cmd/runner.createLimiter({_, _}, {0x0, {0xac9aaf, 0x7}, 0x1f90, {0xad2666, 0x7}, 0x17b6, {0xad111f, ...}, ...}, ...)
	/Users/renuka/git/ratelimit/src/service_cmd/runner/runner.go:55 +0x277
github.com/envoyproxy/ratelimit/src/service_cmd/runner.(*Runner).Run(0xc00038aa80)
	/Users/renuka/git/ratelimit/src/service_cmd/runner/runner.go:121 +0x595
main.main()
	/Users/renuka/git/ratelimit/src/service_cmd/main.go:10 +0x125
```

```sh
$ k get po
my-redis-redis-cluster-0        1/1     Running            0             4m40s
my-redis-redis-cluster-1        1/1     Running            0             4m39s
my-redis-redis-cluster-3        1/1     Running            0             4m39s
my-redis-redis-cluster-2        1/1     Running            0             4m39s
my-redis-redis-cluster-4        1/1     Running            0             4m39s
my-redis-redis-cluster-5        1/1     Running            0             4m39s
ratelimit-5759c7f8d4-p6bs7      0/1     CrashLoopBackOff   5 (37s ago)   3m36s
```

**Comment by renuka-fernando** [comment]
I am getting this even for older rate limit services versions, tried this with docker image `ratelimit-5d54c6d697-c98w8`
Then it looks like a configuration issue.

**Comment by renuka-fernando** [comment]
Ohh could find the issue, rate limit service is not supporting Redis 7.x.x in cluster mode, working on Redis 6.x.x. Tried with the docker image `redis:6.2.7-alpine3.16`.

**Comment by renuka-fernando** [comment]
Tested with the latest mediocregopher/radix/tree/v3: https://github.com/mediocregopher/radix/tree/v3.8.1
It is working with Redis 7 in cluster mode
Looks like it is fixed in 3.8.1 with https://github.com/mediocregopher/radix/pull/322
